### DISCUSSION TITLE
Should I learn DSA in Python or C?

### DISCUSSION THREAD

**Discussion by Rough4993** [community/community#68104] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

I am a first year student of B.Tech (CSE), Should I learn DSA in Python or C .
Help me ???

**Comment by Durgesh4993** [comment]
From the three languages I would go with C++. This is because it gives low level access of C and it also offers high level features through object oriented programming. C provides a platform to learn all low level data structures. In fact, operating system kernels are written in C. I don’t see a better language to learn low level data structures. By learning low level data structures high level data structures are very easy. The good thing about C++ is that most C code is valid C++ code. C++ adds object orientation to C. In fact, C++ was called C with classes during its early days. Java is a high level pure object oriented programming language. It runs on a virtual machine which means it denies you the ability to write code that executes directly on the processor. This means that data structures like [linked lists](https://en.wikipedia.org/wiki/Linked_list) written in Java are not using actual memory references like in C or C++. This denies the learner an opportunity to learn the underlying system in a direct manner. Another thing that I can mention is on the part of [memory management](https://en.wikipedia.org/wiki/Memory_management). Java uses automatic memory management so the programmer does not have to worry about [memory leaks](https://en.wikipedia.org/wiki/Memory_leak). This denies the learner an opportunity to learn about memory management unlike in C or C++ where d[ynamically allocated memory(heap)](https://en.wikipedia.org/wiki/Memory_management#HEAP) must be handled by the programmer. From all these explanations it is clear why I would choose C++ for learning of data structures.

**Reply by emranpy** [reply]
As a first-year student in B.Tech (CSE), learning Data Structures and Algorithms (DSA) is an essential part of your curriculum. When it comes to choosing a programming language for DSA, both Python and C have their own advantages. However, Python is often recommended for beginners for the following reasons:

Simplicity and Readability: Python has a clean and easy-to-understand syntax, making it more readable than C. It uses fewer lines of code to implement algorithms, allowing you to focus more on the logic of DSA concepts rather than dealing with complex language structures.

Rapid Prototyping and Faster Development: Python is known for its fast development cycle. It provides built-in data structures and high-level abstractions that simplify the implementation of DSA algorithms. This allows you to quickly prototype and experiment with different algorithms, which is beneficial when learning and exploring various concepts.

Vast Ecosystem and Libraries: Python has a vast ecosystem with numerous libraries and frameworks that provide ready-to-use implementations of many DSA algorithms. Libraries like NumPy, Pandas, and matplotlib offer powerful tools for data manipulation, analysis, and visualization, which are often required in DSA tasks.

Community Support and Learning Resources: Python has a large and active community of developers, which means you can find plenty of learning resources, tutorials, and online communities to support your DSA journey. This makes it easier to seek help, collaborate with others, and gain insights from experienced programmers.

Career Relevance: Python is widely used in the industry for various applications, including data science, machine learning, web development, and scripting. By learning DSA in Python, you not only gain proficiency in DSA but also acquire transferable skills that are in high demand in the job market.

It's worth noting that learning C for DSA can also be beneficial, as it provides a deeper understanding of low-level memory management and system-level programming. However, for beginners, Python's simplicity, readability, and extensive libraries make it an excellent choice to grasp the fundamental concepts of DSA effectively.

**Comment by Durgesh4993** [comment]
@Rough4993 Hope you like the suggestion

**Comment by Rough4993** [comment]
@Durgesh4993 Mean i should stard DSA in C++

**Comment by Durgesh4993** [comment]
It totally denpend on you

**Comment by Durgesh4993** [comment]
you can also choose Java, it is also help at industry level

**Comment by Rough4993** [comment]
Can i choose python

**Comment by Durgesh4993** [comment]
yes but some time, python did't allow some companies

**Comment by Rough4993** [comment]
ohhh

**Comment by Rough4993** [comment]
But i think Python is easy to learn and fast

**Comment by Durgesh4993** [comment]
Yes @Rough4993  you are right, As i say some companies did't allow to solve DSA question in java as a fresher

**Comment by Durgesh4993** [comment]
Ok

**Comment by Durgesh4993** [comment]
Hope you understand

**Comment by Rough4993** [comment]
Yes, Thank you

**Comment by queenofcorgis** [comment]
GitHub's [Acceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-acceptable-use-policies) prohibits coordinated or inauthentic activity like rapid questions and answers. As a result, we'll be unmarking the answer and locking this post. Please note any future violations may result in a temporary or indefinite block from the Community. Thanks for understanding.
